// @cycleplatform/api-client-typescript/dist/index.js
const Q = /\{[^{}]+\}/g, _ = () => {
  var r, t;
  return typeof process == "object" && Number.parseInt((t = (r = process == null ? void 0 : process.versions) == null ? void 0 : r.node) == null ? void 0 : t.substring(0, 2)) >= 18 && process.versions.undici;
};
function X() {
  return Math.random().toString(36).slice(2, 11);
}
function Z(r) {
  let {
    baseUrl: t = "",
    Request: e = globalThis.Request,
    fetch: o = globalThis.fetch,
    querySerializer: n,
    bodySerializer: s,
    headers: i,
    requestInitExt: l = void 0,
    ...h
  } = { ...r };
  l = _() ? l : void 0, t = L(t);
  const g = [];
  async function p(c, a) {
    const {
      baseUrl: b,
      fetch: I = o,
      Request: k = e,
      headers: O,
      params: R = {},
      parseAs: A = "json",
      querySerializer: E,
      bodySerializer: D = s ?? V,
      body: H,
      middleware: N = [],
      ...x
    } = a || {};
    let C = t;
    b && (C = L(b) ?? t);
    let q = typeof n == "function" ? n : P(n);
    E && (q = typeof E == "function" ? E : P({
      ...typeof n == "object" ? n : {},
      ...E
    }));
    const z = H === void 0 ? void 0 : D(
      H,
      // Note: we declare mergeHeaders() both here and below because it’s a bit of a chicken-or-egg situation:
      // bodySerializer() needs all headers so we aren’t dropping ones set by the user, however,
      // the result of this ALSO sets the lowest-priority content-type header. So we re-merge below,
      // setting the content-type at the very beginning to be overwritten.
      // Lastly, based on the way headers work, it’s not a simple “present-or-not” check becauase null intentionally un-sets headers.
      M(i, O, R.header)
    ), W = M(
      // with no body, we should not to set Content-Type
      z === void 0 || // if serialized body is FormData; browser will correctly set Content-Type & boundary expression
      z instanceof FormData ? {} : {
        "Content-Type": "application/json"
      },
      i,
      O,
      R.header
    ), y = [...g, ...N], G = {
      redirect: "follow",
      ...h,
      ...x,
      body: z,
      headers: W
    };
    let T, v, m = new k(
      Y(c, { baseUrl: C, params: R, querySerializer: q }),
      G
    ), f;
    for (const d in x)
      d in m || (m[d] = x[d]);
    if (y.length) {
      T = X(), v = Object.freeze({
        baseUrl: C,
        fetch: I,
        parseAs: A,
        querySerializer: q,
        bodySerializer: D
      });
      for (const d of y)
        if (d && typeof d == "object" && typeof d.onRequest == "function") {
          const u = await d.onRequest({
            request: m,
            schemaPath: c,
            params: R,
            options: v,
            id: T
          });
          if (u)
            if (u instanceof k)
              m = u;
            else if (u instanceof Response) {
              f = u;
              break;
            } else
              throw new Error("onRequest: must return new Request() or Response() when modifying the request");
        }
    }
    if (!f) {
      try {
        f = await I(m, l);
      } catch (d) {
        let u = d;
        if (y.length)
          for (let w = y.length - 1; w >= 0; w--) {
            const S = y[w];
            if (S && typeof S == "object" && typeof S.onError == "function") {
              const j = await S.onError({
                request: m,
                error: u,
                schemaPath: c,
                params: R,
                options: v,
                id: T
              });
              if (j) {
                if (j instanceof Response) {
                  u = void 0, f = j;
                  break;
                }
                if (j instanceof Error) {
                  u = j;
                  continue;
                }
                throw new Error("onError: must return new Response() or instance of Error");
              }
            }
          }
        if (u)
          throw u;
      }
      if (y.length)
        for (let d = y.length - 1; d >= 0; d--) {
          const u = y[d];
          if (u && typeof u == "object" && typeof u.onResponse == "function") {
            const w = await u.onResponse({
              request: m,
              response: f,
              schemaPath: c,
              params: R,
              options: v,
              id: T
            });
            if (w) {
              if (!(w instanceof Response))
                throw new Error("onResponse: must return new Response() when modifying the response");
              f = w;
            }
          }
        }
    }
    if (f.status === 204 || m.method === "HEAD" || f.headers.get("Content-Length") === "0")
      return f.ok ? { data: void 0, response: f } : { error: void 0, response: f };
    if (f.ok)
      return A === "stream" ? { data: f.body, response: f } : { data: await f[A](), response: f };
    let U = await f.text();
    try {
      U = JSON.parse(U);
    } catch {
    }
    return { error: U, response: f };
  }
  return {
    request(c, a, b) {
      return p(a, { ...b, method: c.toUpperCase() });
    },
    /** Call a GET endpoint */
    GET(c, a) {
      return p(c, { ...a, method: "GET" });
    },
    /** Call a PUT endpoint */
    PUT(c, a) {
      return p(c, { ...a, method: "PUT" });
    },
    /** Call a POST endpoint */
    POST(c, a) {
      return p(c, { ...a, method: "POST" });
    },
    /** Call a DELETE endpoint */
    DELETE(c, a) {
      return p(c, { ...a, method: "DELETE" });
    },
    /** Call a OPTIONS endpoint */
    OPTIONS(c, a) {
      return p(c, { ...a, method: "OPTIONS" });
    },
    /** Call a HEAD endpoint */
    HEAD(c, a) {
      return p(c, { ...a, method: "HEAD" });
    },
    /** Call a PATCH endpoint */
    PATCH(c, a) {
      return p(c, { ...a, method: "PATCH" });
    },
    /** Call a TRACE endpoint */
    TRACE(c, a) {
      return p(c, { ...a, method: "TRACE" });
    },
    /** Register middleware */
    use(...c) {
      for (const a of c)
        if (a) {
          if (typeof a != "object" || !("onRequest" in a || "onResponse" in a || "onError" in a))
            throw new Error("Middleware must be an object with one of `onRequest()`, `onResponse() or `onError()`");
          g.push(a);
        }
    },
    /** Unregister middleware */
    eject(...c) {
      for (const a of c) {
        const b = g.indexOf(a);
        b !== -1 && g.splice(b, 1);
      }
    }
  };
}
function $(r, t, e) {
  if (t == null)
    return "";
  if (typeof t == "object")
    throw new Error(
      "Deeply-nested arrays/objects aren’t supported. Provide your own `querySerializer()` to handle these."
    );
  return `${r}=${(e == null ? void 0 : e.allowReserved) === !0 ? t : encodeURIComponent(t)}`;
}
function F(r, t, e) {
  if (!t || typeof t != "object")
    return "";
  const o = [], n = {
    simple: ",",
    label: ".",
    matrix: ";"
  }[e.style] || "&";
  if (e.style !== "deepObject" && e.explode === !1) {
    for (const l in t)
      o.push(l, e.allowReserved === !0 ? t[l] : encodeURIComponent(t[l]));
    const i = o.join(",");
    switch (e.style) {
      case "form":
        return `${r}=${i}`;
      case "label":
        return `.${i}`;
      case "matrix":
        return `;${r}=${i}`;
      default:
        return i;
    }
  }
  for (const i in t) {
    const l = e.style === "deepObject" ? `${r}[${i}]` : i;
    o.push($(l, t[i], e));
  }
  const s = o.join(n);
  return e.style === "label" || e.style === "matrix" ? `${n}${s}` : s;
}
function J(r, t, e) {
  if (!Array.isArray(t))
    return "";
  if (e.explode === !1) {
    const s = { form: ",", spaceDelimited: "%20", pipeDelimited: "|" }[e.style] || ",", i = (e.allowReserved === !0 ? t : t.map((l) => encodeURIComponent(l))).join(s);
    switch (e.style) {
      case "simple":
        return i;
      case "label":
        return `.${i}`;
      case "matrix":
        return `;${r}=${i}`;
      // case "spaceDelimited":
      // case "pipeDelimited":
      default:
        return `${r}=${i}`;
    }
  }
  const o = { simple: ",", label: ".", matrix: ";" }[e.style] || "&", n = [];
  for (const s of t)
    e.style === "simple" || e.style === "label" ? n.push(e.allowReserved === !0 ? s : encodeURIComponent(s)) : n.push($(r, s, e));
  return e.style === "label" || e.style === "matrix" ? `${o}${n.join(o)}` : n.join(o);
}
function P(r) {
  return function(e) {
    const o = [];
    if (e && typeof e == "object")
      for (const n in e) {
        const s = e[n];
        if (s != null) {
          if (Array.isArray(s)) {
            if (s.length === 0)
              continue;
            o.push(
              J(n, s, {
                style: "form",
                explode: !0,
                ...r == null ? void 0 : r.array,
                allowReserved: (r == null ? void 0 : r.allowReserved) || !1
              })
            );
            continue;
          }
          if (typeof s == "object") {
            o.push(
              F(n, s, {
                style: "deepObject",
                explode: !0,
                ...r == null ? void 0 : r.object,
                allowReserved: (r == null ? void 0 : r.allowReserved) || !1
              })
            );
            continue;
          }
          o.push($(n, s, r));
        }
      }
    return o.join("&");
  };
}
function K(r, t) {
  let e = r;
  for (const o of r.match(Q) ?? []) {
    let n = o.substring(1, o.length - 1), s = !1, i = "simple";
    if (n.endsWith("*") && (s = !0, n = n.substring(0, n.length - 1)), n.startsWith(".") ? (i = "label", n = n.substring(1)) : n.startsWith(";") && (i = "matrix", n = n.substring(1)), !t || t[n] === void 0 || t[n] === null)
      continue;
    const l = t[n];
    if (Array.isArray(l)) {
      e = e.replace(o, J(n, l, { style: i, explode: s }));
      continue;
    }
    if (typeof l == "object") {
      e = e.replace(o, F(n, l, { style: i, explode: s }));
      continue;
    }
    if (i === "matrix") {
      e = e.replace(o, `;${$(n, l)}`);
      continue;
    }
    e = e.replace(o, i === "label" ? `.${encodeURIComponent(l)}` : encodeURIComponent(l));
  }
  return e;
}
function V(r, t) {
  return r instanceof FormData ? r : t && (t.get instanceof Function ? t.get("Content-Type") ?? t.get("content-type") : t["Content-Type"] ?? t["content-type"]) === "application/x-www-form-urlencoded" ? new URLSearchParams(r).toString() : JSON.stringify(r);
}
function Y(r, t) {
  var n;
  let e = `${t.baseUrl}${r}`;
  (n = t.params) != null && n.path && (e = K(e, t.params.path));
  let o = t.querySerializer(t.params.query ?? {});
  return o.startsWith("?") && (o = o.substring(1)), o && (e += `?${o}`), e;
}
function M(...r) {
  const t = new Headers();
  for (const e of r) {
    if (!e || typeof e != "object")
      continue;
    const o = e instanceof Headers ? e.entries() : Object.entries(e);
    for (const [n, s] of o)
      if (s === null)
        t.delete(n);
      else if (Array.isArray(s))
        for (const i of s)
          t.append(n, i);
      else s !== void 0 && t.set(n, s);
  }
  return t;
}
function L(r) {
  return r.endsWith("/") ? r.substring(0, r.length - 1) : r;
}
class ee extends EventTarget {
  constructor(t, e, o) {
    super(), this.promise = this.track(t, e, o);
  }
  async track(t, e, o) {
    const n = typeof e == "object" ? e.id : e, s = {
      pollingInterval: 2e3,
      ...o
    };
    let i = 0;
    for (; ; ) {
      const l = await t.GET("/v1/jobs/{jobId}", {
        params: { path: { jobId: n } }
      });
      if (l.error) {
        if (i < 5 && l.error.error.code === "404.job") {
          i++, await this.delay(s.pollingInterval);
          continue;
        }
        throw l.error;
      }
      const h = l.data.data, g = re(h);
      if (this.dispatchEvent(
        new CustomEvent("progress", {
          detail: { ...g, state: h.state.current }
        })
      ), te(h)) {
        if (h.state.current === "error")
          throw h;
        return this.dispatchEvent(
          new CustomEvent("done", {
            detail: h
          })
        ), h;
      }
      await this.delay(s.pollingInterval);
    }
  }
  delay(t = 2e3) {
    return new Promise((e) => setTimeout(e, t));
  }
}
function te(r) {
  return [
    "completed",
    "error",
    "expired"
  ].includes(r.state.current);
}
const B = "0001-01-01T00:00:00Z";
function re(r) {
  if (!r.tasks || r.tasks.length === 0)
    return { total: 0, completed: 0, failed: 0, percent: 100 };
  const t = r.tasks.reduce(
    (e, o) => {
      var s, i, l;
      const n = ((s = o.steps) == null ? void 0 : s.length) || 1;
      if (e.total += n, o.state.current === "pending")
        return e;
      if (!((i = o.steps) != null && i.length)) {
        switch (o.state.current) {
          case "completed":
            e.completed += 1;
            break;
          case "error":
            e.failed += 1;
        }
        return e;
      }
      return (l = o.steps) == null || l.forEach((h) => {
        if (h.completed !== B) {
          e.completed += 1;
          return;
        }
        ["error", "completed"].includes(o.state.current || "") && h.started !== B && (e.failed += 1);
      }), e;
    },
    { total: 0, completed: 0, failed: 0 }
  );
  return {
    ...t,
    percent: t.completed / t.total * 100
  };
}
function ne(r, t, e) {
  return new ee(r, t, e);
}
function oe({
  apiKey: r,
  baseUrl: t = "https://api.cycle.io",
  hubId: e,
  fetch: o
}) {
  const n = Z({
    baseUrl: t,
    fetch: o || fetch
  }), s = {
    async onRequest({ request: i }) {
      return i.headers.set("Authorization", `Bearer ${r}`), i.headers.set("X-Hub-Id", e), i;
    }
  };
  return n.use(s), n;
}
export {
  oe as getClient,
  re as getJobProgress,
  ne as trackJob,
  B as zeroTimeString
};
